File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> freq (95, 0);
        int max_count = 0;

        int start = 0;
        int end = 0;
        int size = s.size();
        while (start < size && end < size){
            if ((end == size) || freq[s[end]-' ']){
                if (end-start > max_count) max_count = end-start;
                for(int i = start; i < end; i++) freq[s[i]-' '] = 0;
                start ++;
                end = start;
            }

            freq[s[end]-' '] = 1;
            end ++;
        }

        if (end-start > max_count) max_count = end-start;

        return max_count;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        prev_ind = [-1]*100

        ans = 0
        p1 = -1
        for p2, char in enumerate(s):
            if prev_ind[ord(char)-ord(' ')] > p1:
                p1 = prev_ind[ord(char)-ord(' ')]
            
            prev_ind[ord(char)-ord(' ')] = p2
            ans = max(ans, p2-p1)
        
        return ans
